### PR TITLE
Civicrm Contact element improvements

### PR DESCRIPTION
**Pull request by mglaman**
Overview
----------------------------------------
Bring up the feature parity of the existing contact element.

Opening as a draft to help review changes and identify some fixes which could be spun out. It makes certain things configurable, but they do not exactly improve functionality yet.

### CODE REVIEW

File: includes/wf_crm_admin_form.inc [hunk 1]
```diff
@@ -1793,10 +1793,6 @@ class wf_crm_admin_form {
           continue;
         }
         // Webform uses YAML dump, which dies on FormattableMarkup.
-        $enabled_element['extra'] = array_map(
```
Comment (mglaman) [inline includes/wf_crm_admin_form.inc:1796]: This could be merged in as a separate commit. The `stringifyFormattableMarkup` callable is now recursive.

File: includes/wf_crm_webform_base.inc [hunk 2]
```diff
@@ -247,65 +247,62 @@ abstract class wf_crm_webform_base {
     $filters = wf_crm_search_filters($this->node, $component);
     // Start with the url - that trumps everything.
     $query = \Drupal::request()->query;
-    if ($query->has("cid$c") || ($c == 1 && $query->has('cid'))) {
+    if ($query->has("cid$c") || ($c === 1 && $query->has('cid'))) {
```
Comment (mglaman) [inline includes/wf_crm_webform_base.inc]: Don't strict check `$c`, it's probably a string.

File: includes/wf_crm_webform_base.inc [hunk 3]
```diff
@@ -247,65 +247,62 @@ abstract class wf_crm_webform_base {
...
       else {
-        $dupes_allowed = $component['extra']['dupes_allowed'];
+        $dupes_allowed = $component['#dupes_allowed'];
       }
       foreach ($found as $cid) {
         // Don't pick the same contact twice unless explicitly told to do so
         if (!$dupes_allowed) {
           foreach($this->ent['contact'] as $contact) {
-            if (!empty($contact['id']) && $cid == $contact['id']) {
+            if (!empty($contact['id']) && $cid === $contact['id']) {
```
Comment (mglaman) [inline includes/wf_crm_webform_base.inc]: Undo strict checking

File: src/Plugin/WebformElement/CivicrmContact.php [hunk 4]
```diff
@@ -25,135 +27,386 @@
...
+
   /**
    * {@inheritdoc}
    *
    * @see _webform_render_civicrm_contact()
    */
   public function prepare(array &$element, WebformSubmissionInterface $webform_submission = NULL) {
+    // Webform removes values which equal their defaults but does not populate
+    // they keys.
+    $ensure_keys_have_values = [
```
Comment (mglaman) [inline src/Plugin/WebformElement/CivicrmContact.php:86]: Review. If the `defaultProperties` have empty values these should persist.

**Comment by mglaman** [comment]
@KarinG would you mind testing this for regressions? This should expose more settings, I don't know if it _adds_ features but does some cleanup. I just want to make sure we didn't lose some of the existing functionality since we don't have much for automated testing

File: src/Plugin/WebformElement/CivicrmContact.php (hunk 4 above)
Comment (mglaman) [reply src/Plugin/WebformElement/CivicrmContact.php:86]: I don't think this a blocker, just a general "learning Webform"

**Comment by KarinG** [comment]
Pulling this in now.

**Comment by KarinG** [comment]
I'm getting a 

Error: Unsupported operand types in wf_crm_webform_postprocess->isFieldHiddenByExistingContactSettings() (line 2404 of /var/www/drupal/web/modules/contrib/webform_civicrm/includes/wf_crm_webform_postprocess.inc) #0 /var/www/drupal/web/modules/contrib/webform_civicrm/includes/wf_crm_webform_postprocess.inc(2294): wf_crm_webform_postprocess->isFieldHiddenByExistingContactSettings('contact', '1', 'contact', '1', 'first_name', true)

I enabled the Existing Contact checkbox on the CiviCRM (config page): and then tried to submit the form leaving Existing Contact blank:

image

**My YAML**
flexbox:
  '#type': flexbox
  civicrm_1_contact_1_contact_existing:
    '#type': civicrm_contact
    '#title': 'Existing Contact'
    '#widget': autocomplete
    '#results_display':
      display_name: display_name
    '#default': contact_id
    '#extra':
      search_prompt: '- Choose existing -'
      allow_create: 1
      none_prompt: '+ Create new +'
      widget: hidden
      default: user
      contact_type: individual
    '#form_key': civicrm_1_contact_1_contact_existing
  civicrm_1_contact_1_contact_first_name:
    '#type': textfield
    '#contact_type': individual
    '#form_key': civicrm_1_contact_1_contact_first_name
    '#extra':
      width: 20
    '#title': 'First Name'
  civicrm_1_contact_1_contact_last_name:
    '#type': textfield
    '#contact_type': individual
    '#form_key': civicrm_1_contact_1_contact_last_name
    '#extra':
      width: 20
    '#title': 'Last Name'
flexbox_01:
  '#type': flexbox
  civicrm_1_contact_1_address_street_address:
    '#type': textfield
    '#extra':
      width: 60
    '#form_key': civicrm_1_contact_1_address_street_address
    '#title': 'Street Address'
  civicrm_1_contact_1_address_city:
    '#type': textfield
    '#extra':
      width: 20
    '#form_key': civicrm_1_contact_1_address_city
    '#title': City
flexbox_02:
  '#type': flexbox
  civicrm_1_contact_1_address_postal_code:
    '#type': textfield
    '#title': 'Postal Code'
    '#minlength': '7'
    '#maxlength': '7'
    '#size': '7'
    '#extra':
      width: 7
    '#form_key': civicrm_1_contact_1_address_postal_code
  civicrm_1_contact_1_address_country_id:
    '#type': civicrm_options
    '#title': Country
    '#options':
      1039: Canada
      1152: Netherlands
      1228: 'United States'
    '#extra':
      aslist: 1
    '#form_key': civicrm_1_contact_1_address_country_id
    '#data_type': null
    '#default_option': false
    '#default_value': false
  civicrm_1_contact_1_address_state_province_id:
    '#type': textfield
    '#title': State/Province
    '#extra':
      maxlength: 5
      width: 4
    '#data_type': state_province_abbr
    '#form_key': civicrm_1_contact_1_address_state_province_id

**Comment by mglaman** [comment]
Rebased against 8.x-5.x

**Comment by mglaman** [comment]
@KarinG 🤦‍♂ I know why you're receiving the error. The problem is how nested elements work in D8 compared to D7.

```
$component = $this->node->getElement("civicrm_{$c}_contact_1_contact_existing");
```

This returns null. Even though the element exists. That is because of the `civicrm_1_contact_1_contact_existing` element being nested under `flexbox` in an array tree.

Although `getElement` is supposed to search via a flattened array. I wonder if `$include_children` needs to be passed. This is one thing that broke pagination assumptions.

**Comment by KarinG** [comment]
Ah... makes sense. We ran into that with Contributions - and this is what solved it. Can we do something like that? 

https://github.com/colemanw/webform_civicrm/commit/d43c9f3c14a4032ff9105a5eeebc16092ed9cecb

```
$elements = $webform->getElementsDecodedAndFlattened();
```

**Comment by KarinG** [comment]
Testing this now (with and without flexbox-es) - and no longer getting the Error on submission 👍 

There are a number of Existing Contact scenarios/functionality. Testing the first two (used the most): 

1) **Set default contact from: Current User** -> it works!

image

image

2) **Pass contact id in URL** -> So I'm logged in but want to load contact id 5 as cid1 (instead of logged in user). I can't get the First Name and Last Name of cid1=5 - in either form (with or without flexboxes):

URL = /webform/no_flexboxes?cid1=5

Config for the Existing Contact Element:
image

image

**Comment by KarinG** [comment]
Merging this now - it does get the first bit of functionality in there for the Existing Contact Element (logged in user). Onwards!
